### PR TITLE
[multikueue] Job live status update

### PR DESCRIPTION
**Pull request by trasc**
#### What type of PR is this?

/kind feature

#### What this PR does / why we need it:
Add live status updates for multikueue jobs

#### Which issue(s) this PR fixes:

Relates to #693 

#### Does this PR introduce a user-facing change?

```release-note
Add live status updates for multikueue jobs
```

### CODE REVIEW

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 1]
```diff
@@ -97,30 +120,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+			ctrl.LoggerFrom(watchCtx).V(2).Error(err, "unaable to start the watcher", "kind", kind)
+		}
+	}
+
+	rc.kubeconfig = kubeconfig
+	return nil
+}
+
+func (rc *remoteClient) startWatcher(ctx context.Context, kind string, w multiKueueWatcher) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("watchKind", kind)
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:147]: When doing the watcher for jobs please use a label to filter out jobs not-owner by MultiKueue. The analogous approach as we do for GC. This could be the same label reused. IMO no need for a different label key for Job and workload, because the intention is the same. Same label keys can be generally used on different objects.

**Review by mimowo — COMMENTED** [review]
Initial comment about the watch scope. I think we should have a label to make the filtering fast.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 2]
```diff
@@ -97,30 +120,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
 	}
 	rc.client = remoteClient
 
-	newWatcher, err := rc.client.Watch(watchCtx, &kueue.WorkloadList{})
+	err = rc.startWatcher(watchCtx, kueue.GroupVersion.WithKind("Workload").GroupKind().String(), &workloadKueueWatcher{})
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:123]: Why are we watching workloads, and not Jobs directly? If we are only watching workloads, how do we get status updates on the remote Job status update?

File: test/integration/multikueue/multikueue_test.go [hunk 3]
```diff
@@ -456,6 +456,20 @@ var _ = ginkgo.Describe("Multikueue", func() {
...
 		})
 
+		ginkgo.By("changing the status of the jobset in the worker, updates the manager's jobset status", func() {
+			gomega.Eventually(func(g gomega.Gomega) {
+				createdJobSet := jobset.JobSet{}
+				g.Expect(worker2TestCluster.client.Get(worker2TestCluster.ctx, client.ObjectKeyFromObject(jobSet), &createdJobSet)).To(gomega.Succeed())
+				createdJobSet.Status.Restarts = 10
+				g.Expect(worker2TestCluster.client.Status().Update(worker2TestCluster.ctx, &createdJobSet)).To(gomega.Succeed())
+			}, util.Timeout, util.Interval).Should(gomega.Succeed())
+			gomega.Eventually(func(g gomega.Gomega) {
```
Comment (mimowo) [inline test/integration/multikueue/multikueue_test.go:466]: I don't see yet the flow, could you explain. If in the code we watch for updates on the workloads, how do we get notification about the remote job status update (`createdJobSet.Status.Restarts = 10`)?

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 4]
```diff
@@ -276,7 +277,7 @@ func (a *wlReconciler) reconcileGroup(ctx context.Context, group *wlGroup) error
 			Type:    kueue.WorkloadFinished,
 			Status:  metav1.ConditionTrue,
 			Reason:  remoteFinishedCond.Reason,
-			Message: fmt.Sprintf("From remote %q: %s", remote, remoteFinishedCond.Message),
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/workload.go:279]: Why changing this? it could be useful for users to know immediately which worker is processing the job.

File: test/e2e/multikueue/e2e_test.go [hunk 5]
```diff
@@ -255,7 +255,7 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 						Parallelism: 2,
 						Completions: 2,
 						Image:       "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0",
-						Args:        []string{"1ms"},
+						Args:        []string{"2s"}, // give it the time to be observed Active in the live status update step.
```
Comment (mimowo) [inline test/e2e/multikueue/e2e_test.go:258]: I would prefer to give it 10min, and the just terminate the pods, to avoid any risk for flakes.

File: test/e2e/multikueue/e2e_test.go [hunk 6]
```diff
@@ -218,7 +218,7 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 						Type:    kueue.WorkloadFinished,
 						Status:  metav1.ConditionTrue,
 						Reason:  "JobFinished",
-						Message: `From remote "worker1": Job finished successfully`,
```
Comment (mimowo) [inline test/e2e/multikueue/e2e_test.go:221]: Why changing this? It could be useful for users

File: test/e2e/multikueue/e2e_test.go (hunk 5 above)
Comment (trasc) [reply test/e2e/multikueue/e2e_test.go]: done

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:149]: I see, but this means we may actually get events from another MultiKueue instance sharing worker, or other jobs using the prebuilt-workload. I would suggest to add a follow up to align this and use the origin label later on.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:279]: > The message prefix has very little added value since the remote used will remain recorded in the admission check state.

sgtm, in that case I think we could drop the prefix indeed in that case. However, I would like to make to ensure there is a way users can get to know the worker doing computations. Do we have an assert for this in e2e tests? Maybe extend this test to show the information is surfaced.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 1 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:149]: I'll alingn them in the last pr  I marge (this or GC)

File: test/e2e/multikueue/e2e_test.go (hunk 5 above)
Comment (trasc) [reply test/e2e/multikueue/e2e_test.go]: looks like we have different behaviors for different k8s versions ... , may I should revert

Comment (mimowo) [reply test/e2e/multikueue/e2e_test.go]: can you elaborate, what is the difference, why does it affect us?

Comment (mimowo) [reply test/e2e/multikueue/e2e_test.go]: If needed to unblock / move forward, change to 5s, we have another effort ticketed to get rid of the delays which could be used to clean this up. I prefer 5s for consistency, and 2s may be little on a loaded system.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:125]: Hm, here: `err := rc.startWatcher(watchCtx, kind, watcher, client.HasLabels{constants.PrebuiltWorkloadLabel})`?

Why do we then actually watch both the workload and job? I would imagine we watch the job and update the local status. It is not clear to me why we watch both. Please elaborate / add comment.

File: test/e2e/multikueue/e2e_test.go (hunk 5 above)
Comment (trasc) [reply test/e2e/multikueue/e2e_test.go]: I needed to change a couple of job template params , now it works  just like the Jobs in singlecluster e2e.

The ugly part is that the same setup finishes the successfully in k8s 1.28 and 1.29 but with failure in 1.26 and 1.27.

File: test/e2e/multikueue/e2e_test.go [hunk 7]
```diff
@@ -255,7 +255,9 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 						Parallelism: 2,
 						Completions: 2,
 						Image:       "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0",
-						Args:        []string{"1ms"},
+						// Give it the time to be observed Active in the live status update step.
+						// Delete all the pods once active state is detected.
+						Args: []string{"10s"},
```
Comment (mimowo) [inline test/e2e/multikueue/e2e_test.go:260]: Make it high, like 10min. Two reasons:
1. to clearly indicate the intention that we want to check something before finished
2. to make sure the test would fail if the pods weren't actually deleted

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 1 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:149]: sgtm

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 2 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:125]: So, to confirm IIUC we watch jobs for the live status updates.

Still, we also watch remote workloads to trigger local workload reconciliation. This is needed for example when the workload is finished.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 8]
```diff
@@ -98,30 +122,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+		return err
+	}
+
+	// add a watch for all the adapters implementing multiKueueWatcher
+	for kind, adapter := range adapters {
+		watcher, implementsWatcher := adapter.(multiKueueWatcher)
+		if !implementsWatcher {
+			continue
+		}
+		err := rc.startWatcher(watchCtx, kind, watcher, client.HasLabels{constants.PrebuiltWorkloadLabel})
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:136]: Add temporary TODO to use the origin label once available, so that we don't forget.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 9]
```diff
@@ -98,30 +122,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+		if err != nil {
+			// not being able to setup a watcher is not ideal but we can function
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:138]: I would say this is a hard error, to fail fast. Otherwise it may seem like working but the important feature of live updates will be discovered as broken late.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 10]
```diff
@@ -98,30 +122,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+			// with only thw wl watcher, just log the error.
+			ctrl.LoggerFrom(watchCtx).V(2).Error(err, "unaable to start the watcher", "kind", kind)
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:140]:
```suggestion
			ctrl.LoggerFrom(watchCtx).V(2).Error(err, "unable to start the watcher", "kind", kind)
```

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 11]
```diff
@@ -98,30 +122,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+func (rc *remoteClient) startWatcher(ctx context.Context, kind string, w multiKueueWatcher, listopts ...client.ListOption) error {
+	log := ctrl.LoggerFrom(ctx).WithValues("watchKind", kind)
+	newWatcher, err := rc.client.Watch(ctx, w.GetEmptyList())
+	if err != nil {
+		return err
 	}
 
 	go func() {
+		log.V(3).Info("Starting watch")
+		defer log.V(3).Info("Watch ended")
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:157]:
```suggestion
		log.V(2).Info("Starting watch")
		defer log.V(2).Info("Watch ended")
```

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 12]
```diff
@@ -76,17 +78,39 @@ func newClientWithWatch(kubeconfig []byte, options client.Options) (client.WithW
 	return client.NewWithWatch(restConfig, options)
 }
 
+type multiKueueWatcher interface {
+	// returns an empty list of objets
+	GetEmptyList() client.ObjectList
+	// returns the key of the workload of interest
+	// - the object name for workloads
+	// - the prebuilt workload for job types
+	GetEventsWorkloadKey(runtime.Object) (types.NamespacedName, error)
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:87]:
```suggestion
	GetJobKey(runtime.Object) (types.NamespacedName, error)
```
I think the Events is not helping here. Also, we return Job key.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 13]
```diff
@@ -76,17 +78,39 @@ func newClientWithWatch(kubeconfig []byte, options client.Options) (client.WithW
 	return client.NewWithWatch(restConfig, options)
 }
 
+type multiKueueWatcher interface {
+	// returns an empty list of objets
+	GetEmptyList() client.ObjectList
+	// returns the key of the workload of interest
+	// - the object name for workloads
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:85]: I think for workloads it fails looking at the impl?

File: test/e2e/multikueue/e2e_test.go [hunk 14]
```diff
@@ -287,16 +289,36 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 				}, util.Timeout, util.Interval).Should(gomega.Succeed())
 			})
 
+			ginkgo.By("Waiting for the jobSet to get status updates", func() {
+				gomega.Eventually(func(g gomega.Gomega) {
+					createdJobset := &jobset.JobSet{}
+					g.Expect(k8sManagerClient.Get(ctx, client.ObjectKeyFromObject(jobSet), createdJobset)).To(gomega.Succeed())
+
+					g.Expect(createdJobset.Status.ReplicatedJobsStatus).To(gomega.BeComparableTo([]jobset.ReplicatedJobStatus{
```
Comment (mimowo) [inline test/e2e/multikueue/e2e_test.go:297]: I'm wondering this could be rather hard to maintain so that it does not flake with kubernetes or JobSet changes. 

No strong view, but I think it is enough to check a stable subset of fields, just to prove there was an update. We could say use opts to limit the compare just to these fields (or even a subset of them).

**Review by mimowo — COMMENTED** [review]
First pass, looking good overall, we need to have the label to restrict the watch for Jobs. This could be a follow up, or we wait for the GC PR.

**Comment by tenzen-y** [comment]
> is this for jobs or jobsets?

At first glance, this PR is for both job and jobset.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 2 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:125]: yes

File: test/e2e/multikueue/e2e_test.go (hunk 5 above)
Comment (trasc) [reply test/e2e/multikueue/e2e_test.go]: reverted

File: test/integration/multikueue/multikueue_test.go (hunk 3 above)
Comment (mimowo) [reply test/integration/multikueue/multikueue_test.go:466]: gotcha, we will prefer to use the origin label ultimately

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 13 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:85]: Oh, sorry, I got confused by the same function implemented both for Workloads and Jobs. I looked at the Jobs implementations and thought it would error for Workloads. 

Now I see what you do, though the extra layer of indirection makes it slightly hard to follow, but I see there are trade-offs, and this way you can commonize the code for Workloads and Jobs.

File: pkg/controller/admissionchecks/multikueue/workload.go (hunk 4 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go:279]: I see, it is here in e2e tests: https://github.com/kubernetes-sigs/kueue/blob/a8c11f351858cd159a3f823b15860918c0490cbc/test/e2e/multikueue/e2e_test.go#L208

In that case, I'm ok to revmove it from the condition message.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 15]
```diff
@@ -98,30 +122,63 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
 	}
 
 	go func() {
+		log.V(3).Info("Starting watch")
+		defer log.V(3).Info("Watch ended")
 		for r := range newWatcher.ResultChan() {
-			rc.queueWorkloadEvent(watchCtx, r)
+			wlKey, err := w.GetEventsWorkloadKey(r.Object)
+			if err != nil {
+				log.V(2).Error(err, "Cannot get workload key")
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:163]: Log the namespace/name for the object for which we got the error, so that we can identify it

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 16]
```diff
@@ -98,30 +122,65 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
 	}
 	rc.client = remoteClient
 
-	newWatcher, err := rc.client.Watch(watchCtx, &kueue.WorkloadList{})
+	err = rc.startWatcher(watchCtx, kueue.GroupVersion.WithKind("Workload").GroupKind().String(), &workloadKueueWatcher{})
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:125]: Add the comment here as well to use the origin label when introduced

**Review by mimowo — COMMENTED** [review]
Almost LGTM, just nits + we need to migrate to the origin label.

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 16 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:126]: done

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 15 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:164]: `r.Object` is a runtime object and is not exposing the name, however I made it part of the error message for JobSet.

Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:164]: Ok, it is acceptable if there is no easy way to know the name / namespace of the object here. 

However, if possible I prefer to log it here so that it does not need to be done per adapter. Let see what are the options to make it generic.

Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:164]: I guess we could at least log its gvk in a generic way.
Maybe we could also cast it on a Object containing name / namespace

Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go]: gvk added

File: test/e2e/multikueue/e2e_test.go [hunk 17]
```diff
@@ -255,7 +255,8 @@ var _ = ginkgo.Describe("MultiKueue", func() {
 						Parallelism: 2,
 						Completions: 2,
 						Image:       "gcr.io/k8s-staging-perf-tests/sleep:v0.1.0",
-						Args:        []string{"1ms"},
+						// Give it the time to be observed Active in the live status update step.
+						Args: []string{"5s"},
```
Comment (alculquicondor) [inline test/e2e/multikueue/e2e_test.go:259]: Can it be less? Maybe 3s?

**Review by alculquicondor — COMMENTED** [review]
/approve
/hold
for comments
Leaving LGTM to @mimowo

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 18]
```diff
@@ -98,30 +122,66 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
 	go func() {
+		log.V(2).Info("Starting watch")
+		defer log.V(2).Info("Watch ended")
 		for r := range newWatcher.ResultChan() {
-			rc.queueWorkloadEvent(watchCtx, r)
+			wlKey, err := w.GetWorkloadKey(r.Object)
+			if err != nil {
+				log.V(2).Error(err, "Cannot get workload key", "jobKind", r.Object.GetObjectKind().GroupVersionKind())
+			} else {
+				rc.queueWorkloadEvent(ctx, wlKey)
```
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:169]: In fact it needs to be done in workload.go, but let's keep it as is for now.

File: test/e2e/multikueue/e2e_test.go (hunk 17 above)
Comment (mimowo) [reply test/e2e/multikueue/e2e_test.go:259]: Yeah, I suggested 5s for consistency with other places. Ideally, I would like not to have 10min timeout, and once the checks are done, then finish the job, but this is ticketed under https://github.com/kubernetes-sigs/kueue/issues/1449

File: pkg/controller/admissionchecks/multikueue/workload.go [hunk 19]
```diff
@@ -61,8 +61,9 @@ type wlReconciler struct {
 var _ reconcile.Reconciler = (*wlReconciler)(nil)
 
 type jobAdapter interface {
-	// Creates the Job object in the worker cluster using remote client.
-	CreateRemoteObject(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName, workloadName string) error
+	// Creates the Job object in the worker cluster using remote client, if not already created.
+	// Copy the status from the remote job if already exists.
+	SyncJob(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName, workloadName string) error
 	// Copy the status from the job in the worker cluster to the local one.
 	CopyStatusRemoteObject(ctx context.Context, localClient client.Client, remoteClient client.Client, key types.NamespacedName) error
```
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/workload.go]: nice

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 20]
```diff
@@ -102,30 +125,66 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
 	}
 	rc.client = remoteClient
 
-	newWatcher, err := rc.client.Watch(watchCtx, &kueue.WorkloadList{})
+	// TODO: Filter based on multikueue-origin label when available.
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:128]: Remove the TODO since the code already is using the label

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go [hunk 21]
```diff
@@ -102,30 +125,66 @@ func (rc *remoteClient) setConfig(watchCtx context.Context, kubeconfig []byte) e
...
+		return err
+	}
+
+	// add a watch for all the adapters implementing multiKueueWatcher
+	for kind, adapter := range adapters {
+		watcher, implementsWatcher := adapter.(multiKueueWatcher)
+		if !implementsWatcher {
+			continue
+		}
+		// TODO: Filter based on multikueue-origin label when available.
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/multikueuecluster.go:140]: Remove TODO

File: pkg/controller/admissionchecks/multikueue/batchjob_adapter.go [hunk 22]
```diff
@@ -30,14 +32,41 @@ type batchJobAdapter struct{}
...
+	if client.IgnoreNotFound(err) != nil {
+		return err
+	}
+
+	// the remote job exists
+	if err == nil {
+		// This will no longer be necessary when batchJob will support live status update, by then
+		// we should only sync the Status of the job if it's "Finished".
+		remoteFinished := false
+		for _, c := range remoteJob.Status.Conditions {
```
Comment (mimowo) [inline pkg/controller/admissionchecks/multikueue/batchjob_adapter.go:53]: nit: Please use utility function (IsStatusConditionTrue) instead of looping here

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:169]: IIUC the purpose is to batch multiple workload updates by time? Is the refactoring planned as one of the follow ups?

**Review by mimowo — COMMENTED** [review]
LGTM, just some nits / quesiton

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 20 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go]: done

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 21 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go]: done

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:167]: I guess it could be added in the follow-up list.

File: pkg/controller/admissionchecks/multikueue/batchjob_adapter.go (hunk 22 above)
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/batchjob_adapter.go:53]: these are not metav1.Conditions

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (alculquicondor) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:167]: Yes, please. Add it to the follow up list.

File: test/e2e/multikueue/e2e_test.go (hunk 17 above)
Comment (alculquicondor) [reply test/e2e/multikueue/e2e_test.go:259]: sg

File: pkg/controller/admissionchecks/multikueue/multikueuecluster.go (hunk 18 above)
[the remark this answers is not in the document]
Comment (trasc) [reply pkg/controller/admissionchecks/multikueue/multikueuecluster.go:167]: done

File: pkg/controller/admissionchecks/multikueue/batchjob_adapter.go (hunk 22 above)
Comment (mimowo) [reply pkg/controller/admissionchecks/multikueue/batchjob_adapter.go:53]: sg

**Comment by alculquicondor** [comment]
/release-note-edit
```release-note
MultiKueue: Add live status updates for multikueue JobSets
```
